src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs

```diff
@@ -144,11 +144,21 @@ private static bool TryUseSubstring(ReadOnlySpan<string> uniqueStrings, bool ign
                         break;
                     }
 
-                    // All substrings so far are still ASCII only.  If this one contains any ASCII
-                    // letters, mark that we can't switch to case-sensitive.
-                    if (canSwitchIgnoreCaseHashToCaseSensitive && ContainsAnyLetters(substring))
+                    // All substrings so far are still ASCII only.  Now we check whether
+                    // the entire string either is not ASCII or contains ASCII letters,
+                    // in which case we disable the switch to case sensitive.
+                    if (canSwitchIgnoreCaseHashToCaseSensitive)
                     {
-                        canSwitchIgnoreCaseHashToCaseSensitive = false;
+                        // If we are testing the entire string then s == substring
+                        // and we have already checked IsAllAscii(s).
+                        if (isSubstring && !IsAllAscii(s.AsSpan()))
```

Comment:
The existing naming employed here is confusing in conjunction with this comment, since "is substring" being true means s _is not_ substring.

Let's rename "s" to "uniqueString" and "substring" to "uniqueStringSlice", or something like that.